README.md
# CcValidation

[![build status](https://gitlab.com/mtchavez/ex_cc_validation/badges/master/build.svg)](https://gitlab.com/mtchavez/ex_cc_validation/commits/master)
[![coverage report](https://gitlab.com/mtchavez/ex_cc_validation/badges/master/coverage.svg)](https://gitlab.com/mtchavez/ex_cc_validation/commits/master)

Credit Card format validation

## Installation

If [available in Hex](https://hex.pm/docs/publish), the package can be installed
by adding `cc_validation` to your list of dependencies in `mix.exs`:

```elixir
def deps do
  [{:cc_validation, "~> 0.1.0"}]
end
```

## Usage

### Validate a card number

Valid credit card

```elixir
iex> CcValidation.validate("4716892095589823")
{:ok, true}
````

Invalid credit card

```elixir
iex> CcValidation.validate("4024007106963124")
{:error, false}
````

Test card returns false

```elixir
iex> CcValidation.validate("4111111111111111")
{:error, false}
````

### Validate test card numbers when check_test_numbers is true

A valid test card

```elixir
iex> CcValidation.validate("4111111111111111", true)
{:ok, true, test_number: true}
````

An invalid test card

```elixir
iex> CcValidation.validate("4212121212121212", true)
{:error, false, test_number: false}
````

A valid card when checking for test numbers just passes through
to the validate function but will also return that it is not
a test card

```elixir
iex> CcValidation.validate("4716892095589823", true)
{:ok, true, test_number: false}
````

### Card numbers that are of length less than 13 or greater than 19 are invalid.

```elixir
iex> CcValidation.validate("123")
{:error, false}

iex> CcValidation.validate("12345678901234567890")
{:error, false}
````

## Tests

Tests can be ran with `mix` by running `mix test`

## Documentation

Documentation can be generated with [ExDoc](https://github.com/elixir-lang/ex_doc)
and published on [HexDocs](https://hexdocs.pm). Once published, the docs can
be found at [https://hexdocs.pm/cc_validation](https://hexdocs.pm/cc_validation).



CHANGELOG.md
## 0.1.1 - 2017-07-13

- Moved project to Gitlab
- Changed to GitlabCI instead of TravisCI
- Use Dogma for linting and fix linting warnings/errors

## 0.1.0 - 2017-06-24

Initial release

- Validate credit card numbers
- Can check for test card numbers


test/cc_validation_test.exs
defmodule CcValidationTest do
  use ExUnit.Case
  doctest CcValidation
end


test/ex_cc_validation_test.exs
defmodule CcValidationTest do
  use ExUnit.Case
  doctest CcValidation

  test "the truth" do
    assert 1 + 1 == 2
  end
end


lib/cc_validation/test_numbers.ex
defmodule CcValidation.TestNumbers do
  @moduledoc """
  CcValidation.TestNumbers contains test card numbers to validate against
  as a valid card number
  """

  @numbers %{
    "378282246310005": "American Express",
    "371449635398431": "American Express",
    "378734493671000": "American Express Corporate",
    "5610591081018250": "Australian BankCard",
    "30569309025904": "Diners Club",
    "38520000023237": "Diners Club",
    "6011111111111117": "Discover",
    "6011000990139424": "Discover",
    "3530111333300000": "JCB",
    "3566002020360505": "JCB",
    "5555555555554444": "MasterCard",
    "5105105105105100": "MasterCard",
    "4111111111111111": "Visa",
    "4012888888881881": "Visa",
    "4222222222222": "Visa"
  }


  @doc """
  Check that a number is a valid test number or not

  A valid test card
      iex> CcValidation.TestNumbers.exists?("4111111111111111")
      true

  An invalid test card
      iex> CcValidation.TestNumbers.exists?("4212121212121212")
      false
  """
  def exists?(number) do
    Map.has_key? @numbers, String.to_atom(number)
  end
end


mix.exs
defmodule CcValidation.Mixfile do
  use Mix.Project

  def project do
    [
      app: :cc_validation,
      version: "0.1.1",
      elixir: "~> 1.4",
      build_embedded: Mix.env == :prod,
      start_permanent: Mix.env == :prod,
      deps: deps(),
      description: description(),
      package: package(),

      # Test
      test_coverage: [tool: ExCoveralls],
      preferred_cli_env: [
        "coveralls": :test,
        "coveralls.detail": :test,
        "coveralls.post": :test,
        "coveralls.html": :test
      ],


      # Docs
      name: "CcValidation",
      source_url: "https://gitlab.com/mtchavez/ex_cc_validation",
      homepage_url: "https://gitlab.com/mtchavez/ex_cc_validation",
      docs: [
        main: "CcValidation", # The main page in the docs
        extras: ["README.md"]
      ]
  ]
  end

  # Configuration for the OTP application
  #
  # Type "mix help compile.app" for more information
  def application do
    # Specify extra applications you'll use from Erlang/Elixir
    [extra_applications: [:logger]]
  end

  # Dependencies can be Hex packages:
  #
  #   {:my_dep, "~> 0.3.0"}
  #
  # Or git/path repositories:
  #
  #   {:my_dep, git: "https://github.com/elixir-lang/my_dep.git", tag: "0.1.0"}
  #
  # Type "mix help deps" for more examples and options
  defp deps do
    [
      # Dev
      {:dogma, "~> 0.1", only: :dev},
      {:ex_doc, "~> 0.14", only: :dev, runtime: false},

      # Test
      {:excoveralls, "~> 0.7", only: :test}
    ]
  end

  defp description do
    """
    Validate credit card number format using Luhn Algorithm and allows for
    checking test card numbers.
    """
  end

  defp package do
    # These are the default files included in the package
    [
      name: :cc_validation,
      files: ["lib", "mix.exs", "README*", "LICENSE*"],
      maintainers: ["mtchavez"],
      licenses: ["MIT"],
      links: %{
        "Project Source Code" => "https://gitlab.com/mtchavez/ex_cc_validation"
      }
    ]
  end
end


config/config.exs
# This file is responsible for configuring your application
# and its dependencies with the aid of the Mix.Config module.
use Mix.Config

# This configuration is loaded before any dependency and is restricted
# to this project. If another project depends on this project, this
# file won't be loaded nor affect the parent project. For this reason,
# if you want to provide default values for your application for
# 3rd-party users, it should be done in your "mix.exs" file.

# You can configure for your application as:
#
#     config :cc_validation, key: :value
#
# And access this configuration in your application as:
#
#     Application.get_env(:cc_validation, :key)
#
# Or configure a 3rd-party app:
#
#     config :logger, level: :info
#

# It is also possible to import configuration files, relative to this
# directory. For example, you can emulate configuration per environment
# by uncommenting the line below and defining dev.exs, test.exs and such.
# Configuration from the imported file will override the ones defined
# here (which is why it is important to import them last).
#
#     import_config "#{Mix.env}.exs"


lib/cc_validation.ex
defmodule CcValidation do
  @moduledoc """
  CcValidation implements methods to verify validity of a credit card number
  """

  @doc """
  Validate a credit card number using Luhn Algorithm

  ## Examples

  Valid credit card
  iex> CcValidation.validate("4716892095589823")
  {:ok, true}

  Invalid credit card
  iex> CcValidation.validate("4024007106963124")
  {:error, false}

  Test card returns false
  iex> CcValidation.validate("4111111111111111")
  {:error, false}

  ### Validate test card numbers when check_test_numbers is true

  A valid test card
  iex> CcValidation.validate("4111111111111111", true)
  {:ok, true, test_number: true}

  An invalid test card
  iex> CcValidation.validate("4212121212121212", true)
  {:error, false, test_number: false}

  A valid card when checking for test numbers just passes through
  to the validate function but will also return that it is not
  a test card
  iex> CcValidation.validate("4716892095589823", true)
  {:ok, true, test_number: false}

  ### Card numbers that are of length less than 13 or greater than 19 are
  invalid.

  iex> CcValidation.validate("123")
  {:error, false}

  iex> CcValidation.validate("12345678901234567890")
  {:error, false}
  """
  def validate(number, check_test_numbers \\ false)
  def validate(number, true) do
    case CcValidation.TestNumbers.exists?(number) do
      true ->
        {:ok, true, test_number: true}
      false ->
        Tuple.append(validate(number), test_number: false)
    end
  end
  def validate(number, false)
  when byte_size(number) < 13 or byte_size(number) > 19 do
    {:error, false}
  end
  def validate(number, false) do
    case CcValidation.TestNumbers.exists?(number) do
      true ->
        {:error, false}
      _ ->
        total = number
        |> String.codepoints
        |> Enum.reverse
        |> Stream.with_index
        |> Enum.map(fn {num, i} ->
          case rem(i + 1, 2) do
            0 ->
              doubled = String.to_integer(num) * 2
              if doubled > 9 do
                doubled - 9
              else
                doubled
              end
            _ ->
              String.to_integer(num)
          end
        end)
        |> Enum.sum

        case rem(total, 10) do
          0 ->
            {:ok, true}
          _ ->
            {:error, false}
        end
    end
  end
end
